Adding rates_calculator.ipynb

AlessMor opened the pull request:
The notebook can estimate the startup time and net power loss of a DD reactor.

The reactivities are taken from cfspopcon and reported at the beginning of the file for convenience.

The volumetric_reaction_rates_DDfuel function is used to calculate the volumetric reaction rates for DDn, DDp, DT and DHe3 fusion channels inside a DD mixture.

AlessMor committed:
rates_calculator.ipynb uploaded

AlessMor committed:
TBR added

AlessMor committed:
revised equations for T_dot_tot

RemDelaporteMathurin reviewed (commented):
Hi @AlessMor thanks for this!

As a first pass, could you

- [ ] move the functions to a python script that is then imported in the notebook for computation
- [ ] consider using [`pint` for unit conversion](https://pint.readthedocs.io/en/stable/)

I'll try to give this a closer look when I have a bit of time

AlessMor committed:
divided functions in py scripts + use of pint

AlessMor commented:
Hi @RemDelaporteMathurin ,

I've implemented the use of pint and moved the functions in 3 python scripts:
- **sigmav_functions.py** : collection of all sigmav functions from cfspopcon + pint unit
- **volumetric_reaction_rates.py** : contains a single function returning a dictionary of all the volumetric reaction rates for a DD mixture
- **power.py** : contains the functions to calculate the fusion power of a 50D-50T mixture and the net electric power (given Q, eta)

Since I'm using a ipynb main file, I've decided to create a single **units_and_constants.py** file to be called by the main file, as well as the other functions

AlessMor committed:
calculation of P_e_net now can be done either by using Q or P_aux

At the moment Paux is set in order to evaluate P_e_net

AlessMor committed:
parametric_analysis.ipynb file added

calculate_reaction_rates now works also with density and temperature profiles

AlessMor committed:
Sobol's indices calculated in parametric_analysis.ipynb

AlessMor committed:
added sobol_indices.ipynb

sobol_indices.ipynb can be used to calculate the Sobol indices associated to all calculated quantities ($_lost, t_startup,...), by setting the ranges for every free parameter ("V_plasma", "tau_p_T", "tau_p_He3", "n_e_avg", "T_e_avg", "I_ST", "TBR", "TBR_DDn", "P_aux", "eta_th", "Cost_per_kWh")

AlessMor committed:
included tritium decay and main_with_profiles

now all scripts take into account tritium decay when calculating the startup time (be careful, this may lead to infinite time to reach satrtup inventory if conditions are not favourable).
A new script "main_with_profiles.ipynb" has been added, that considers profiles of T and n in order to account for increased efficiency in the centre of the core.
The use of profiles hasn't been implemented in the parametric analysis, nor in the sobol indices yet.

AlessMor committed:
startup_time_plot.ipynb added, plus minor fixes

startup_time_plot.ipynb allows to see the impact of startup inventory and tritium production rate on the startup time, taking into account decay.

SamueleMeschini committed:
improve formatting

SamueleMeschini committed:
Merge branch 'main' of https://github.com/AlessMor/dd_startup into main

improve formatting

SamueleMeschini committed:
add pairplot

SamueleMeschini committed:
add KDE plot

RemDelaporteMathurin commented:
@AlessMor @SamueleMeschini I see this is growing and growing! Should we schedule a call to review all this together and then merge it if we're happy so that we can have future more dedicated PRs?

AlessMor commented:
Thank you @RemDelaporteMathurin ! Since Samuele and I will be in Boston for SOFE, perhaps we can discuss the current status of the project in person. Would that work for you?
If needed, we could schedule a follow-up call after the conference to finalize the merge.

RemDelaporteMathurin commented:
Sounds perfect!
